File: src/harmonica/mapping.ts
import type { HarmonicaCandidate, HarmonicaKey } from "../music/types";

/** Middle C is used as the provisional 1. Replace this after in-game testing. */
export const ROOT_MIDI = 60;

interface BaseKey {
  key: HarmonicaKey;
  degree: 1 | 2 | 3 | 4 | 5 | 6 | 7;
  semitone: number;
  intrinsicOctave: number;
  keyIndex: number;
}

export const BASE_KEYS: readonly BaseKey[] = [
  { key: "Z", degree: 1, semitone: 0, intrinsicOctave: 0, keyIndex: 0 },
  { key: "X", degree: 2, semitone: 2, intrinsicOctave: 0, keyIndex: 1 },
  { key: "C", degree: 3, semitone: 4, intrinsicOctave: 0, keyIndex: 2 },
  { key: "V", degree: 4, semitone: 5, intrinsicOctave: 0, keyIndex: 3 },
  { key: "B", degree: 5, semitone: 7, intrinsicOctave: 0, keyIndex: 4 },
  { key: "N", degree: 6, semitone: 9, intrinsicOctave: 0, keyIndex: 5 },
  { key: "M", degree: 7, semitone: 11, intrinsicOctave: 0, keyIndex: 6 },
  { key: ",", degree: 1, semitone: 12, intrinsicOctave: 1, keyIndex: 7 }
] as const;

export const MAPPING_ASSUMPTIONS = [
  "Z X C V B N M , 对应 1 2 3 4 5 6 7 高音1",
  "升调 = 鼠标右键，降调 = 鼠标左键，半音 = 鼠标中键",
  "升调 / 降调暂按 ±1 八度处理，半音暂按升半音（#）处理",
  "暂定八度修饰与半音修饰可同时按住"
] as const;

const OCTAVE_MODIFIERS = [-1, 0, 1] as const;

export function candidatesForPitch(pitch: number): HarmonicaCandidate[] {
  const candidates: HarmonicaCandidate[] = [];

  for (const base of BASE_KEYS) {
    for (const octaveModifier of OCTAVE_MODIFIERS) {
      for (const sharp of [false, true] as const) {
        const resultingPitch = ROOT_MIDI + base.semitone + octaveModifier * 12 + (sharp ? 1 : 0);
        if (resultingPitch !== pitch) continue;

        candidates.push({
          key: base.key,
          degree: base.degree,
          keyIndex: base.keyIndex,
          intrinsicOctave: base.intrinsicOctave,
          octaveModifier,
          sharp,
          pitch: resultingPitch
        });
      }
    }
  }

  return candidates;
}

export function displayOctave(candidate: HarmonicaCandidate): number {
  return candidate.intrinsicOctave + candidate.octaveModifier;
}

export function midiName(pitch: number): string {
  const names = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"];
  return `${names[pitch % 12]}${Math.floor(pitch / 12) - 1}`;
}

File: desktop/DFH.Core/Harmonica/Mapping.cs
using DFH.Core.Music;

namespace DFH.Core.Harmonica;

/// <summary>One way to sound a pitch: a note key plus octave / sharp modifiers.</summary>
public sealed record HarmonicaCandidate(
    string Key,
    int Degree,
    int KeyIndex,
    int IntrinsicOctave,
    int OctaveModifier,
    bool Sharp,
    int Pitch);

/// <summary>A note together with the fingering the optimizer chose for it.</summary>
public sealed record GameNote
{
    public required NoteEvent Note { get; init; }
    public required HarmonicaCandidate Candidate { get; init; }
    /// <summary>Position of this note in the list handed to the optimizer.</summary>
    public required int SourceIndex { get; init; }

    public int Pitch => Candidate.Pitch;
    public double Start => Note.Start;
    public double Duration => Note.Duration;
    public double? Beat => Note.Beat;
    public double? DurationBeats => Note.DurationBeats;
    public string Key => Candidate.Key;
    public int Degree => Candidate.Degree;
    public int KeyIndex => Candidate.KeyIndex;
    public int OctaveModifier => Candidate.OctaveModifier;
    public bool Sharp => Candidate.Sharp;
}

public sealed record BaseKey(string Key, int Degree, int Semitone, int IntrinsicOctave, int KeyIndex);

/// <summary>Port of src/harmonica/mapping.ts.</summary>
public static class Mapping
{
    /// <summary>Middle C is the provisional 1.</summary>
    public const int RootMidi = 60;

    public static readonly IReadOnlyList<BaseKey> BaseKeys =
    [
        new("Z", 1, 0, 0, 0),
        new("X", 2, 2, 0, 1),
        new("C", 3, 4, 0, 2),
        new("V", 4, 5, 0, 3),
        new("B", 5, 7, 0, 4),
        new("N", 6, 9, 0, 5),
        new("M", 7, 11, 0, 6),
        new(",", 1, 12, 1, 7)
    ];

    private static readonly int[] OctaveModifiers = [-1, 0, 1];

    /// <summary>Candidates in the exact order the web enumerates them; the optimizer's tie-breaking depends on it.</summary>
    public static List<HarmonicaCandidate> CandidatesForPitch(int pitch)
    {
        var candidates = new List<HarmonicaCandidate>();
        foreach (var baseKey in BaseKeys)
        {
            foreach (var octaveModifier in OctaveModifiers)
            {
                foreach (var sharp in new[] { false, true })
                {
                    var resulting = RootMidi + baseKey.Semitone + octaveModifier * 12 + (sharp ? 1 : 0);
                    if (resulting != pitch) continue;
                    candidates.Add(new HarmonicaCandidate(baseKey.Key, baseKey.Degree, baseKey.KeyIndex, baseKey.IntrinsicOctave, octaveModifier, sharp, resulting));
                }
            }
        }
        return candidates;
    }

    public static int DisplayOctave(HarmonicaCandidate candidate) => candidate.IntrinsicOctave + candidate.OctaveModifier;

    /// <summary>↑ / ↓ for the octave mouse buttons, # for the semitone button.</summary>
    public static string ModifierLabel(HarmonicaCandidate candidate)
    {
        var octave = candidate.OctaveModifier > 0 ? "↑" : candidate.OctaveModifier < 0 ? "↓" : "";
        return octave + (candidate.Sharp ? "#" : "");
    }

    private static readonly string[] NoteNames = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"];

    public static string MidiName(int pitch) => $"{NoteNames[((pitch % 12) + 12) % 12]}{(int)Math.Floor(pitch / 12.0) - 1}";
}
